{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CMM262 Programming Practice Problems in Python, Part 3\n",
    "\n",
    "**Author:** Michelle Franc Ragsac (mragsac@eng.ucsd.edu)\n",
    "\n",
    "This notebook contains more exercises that incorporate basic concepts of programming, specifically for the Python programming language, but with a bioinformatics twist! \n",
    "\n",
    "These exercises are quite involved and can be difficult as they require you to read in files from our `ref/` folder, while also manipulating them into data structures in Python.  \n",
    "\n",
    "> * Many of these exercises (and more!) can be found through the Rosalind \"Bioinformatics Stronghold\" resource: http://rosalind.info/problems/list-view/?location=bioinformatics-stronghold"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "### [Calculating Protein Mass](http://rosalind.info/problems/prtm/)\n",
    "\n",
    "<div class=\"alert alert-block alert-success\">\n",
    "    <p><b>Exercise:</b></p>\n",
    "    <p><b><i>Given</i></b>: A protein string, <code>p</code>, of at most 1000 aa\n",
    "    <br><b><i>Return</i></b>: The total weight of <code>p</code></p>\n",
    "    <p><i>Hint:</i> We have a monoisotopic mass table that you can reference: <code>ref/monoisotopic_masses.txt</code>!</p>\n",
    "</div>\n",
    "\n",
    "#### Sample Dataset\n",
    "\n",
    "```\n",
    "SKADYEK\n",
    "```\n",
    "\n",
    "#### Sample Output\n",
    "\n",
    "```\n",
    "821.392\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total weight of the protein string [ SKADYEK ] is : 821.392\n"
     ]
    }
   ],
   "source": [
    "# 1. Initialize the protein string that we want to evaluate\n",
    "p = \"SKADYEK\"\n",
    "\n",
    "################################################################################\n",
    "# 2. Import the monoisotopic mass table and convert it to a dictionary\n",
    "mass_table_filename = \"ref/monoisotopic_masses.txt\"\n",
    "\n",
    "masses = {}\n",
    "with open(mass_table_filename, \"r\") as file:\n",
    "    for line in file:\n",
    "        amino_acid,mass = line.strip().split()\n",
    "        masses[amino_acid.upper()] = float(mass)\n",
    "        \n",
    "################################################################################\n",
    "# 3. Calculate the mass of the protein string\n",
    "total_mass = sum([masses[aa] for aa in p])\n",
    "total_mass = round(total_mass, 3) # round the answer to three decimal places\n",
    "\n",
    "# Share the output\n",
    "print(f\"The total weight of the protein string [ {p} ] is : {total_mass}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "### [Translating RNA into Protein](http://rosalind.info/problems/prot/)\n",
    "\n",
    "<div class=\"alert alert-block alert-success\">\n",
    "    <p><b>Exercise:</b></p>\n",
    "    <p><b><i>Given</i></b>: A RNA string, <code>s</code>, corresponding to a strand of mRNA (of length at most 10 kbp)\n",
    "    <br><b><i>Return</i></b>: The protein string encoded by <code>s</code></p>\n",
    "    <p><i>Hint:</i> We have a RNA codon table that you can reference: <code>ref/rna_codon_table.txt</code>!</p>\n",
    "</div>\n",
    "\n",
    "#### Sample Dataset\n",
    "\n",
    "```\n",
    "AUGGCCAUGGCGCCCAGAACUGAGAUCAAUAGUACCCGUAUUAACGGGUGA\n",
    "```\n",
    "\n",
    "#### Sample Output\n",
    "\n",
    "```\n",
    "MAMAPRTEINSTRING\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The resulting output for this exercise:\n",
      "\n",
      "RNA String:\tAUGGCCAUGGCGCCCAGAACUGAGAUCAAUAGUACCCGUAUUAACGGGUGA\n",
      "Protein String:\tMAMAPRTEINSTRING\n"
     ]
    }
   ],
   "source": [
    "# 1. Initialize the RNA string that we want to convert to a protein string\n",
    "s = \"AUGGCCAUGGCGCCCAGAACUGAGAUCAAUAGUACCCGUAUUAACGGGUGA\"\n",
    "\n",
    "################################################################################\n",
    "# 2. Import the RNA codon table and convert it to a dictionary \n",
    "rna_codon_table_filename = \"ref/rna_codon_table.txt\"\n",
    "\n",
    "codon_table = {}\n",
    "with open(rna_codon_table_filename, \"r\") as file:\n",
    "    for line in file:\n",
    "        line = line.strip().split()\n",
    "        for codon,aa in zip([line[e] for e in range(len(line)) if e % 2 == 0],\n",
    "                            [line[e] for e in range(len(line)) if e % 2 != 0]):\n",
    "            codon_table[codon.upper()] = aa.upper() # set uniform case\n",
    "            \n",
    "################################################################################\n",
    "# 3. Determine the codons we need to evaluate in the RNA string and then\n",
    "#    figure out what the results are! \n",
    "rna_codons = [s[i:i+3].upper() for i in range(0, len(s), 3)]\n",
    "\n",
    "# We also consider if a STOP codon appears earlier than expected and only gather \n",
    "# the portion of the string that we're interested in\n",
    "protein_string = \"\".join([codon_table[codon] for codon in rna_codons])\n",
    "protein_string = protein_string.split(\"STOP\")[0]\n",
    "\n",
    "# Share the output\n",
    "print(\"The resulting output for this exercise:\\n\")\n",
    "print(f\"RNA String:\\t{s}\\nProtein String:\\t{protein_string}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "### [Consensus and Profile](http://rosalind.info/problems/cons/)\n",
    "\n",
    "<div class=\"alert alert-block alert-success\">\n",
    "    <p><b>Exercise:</b></p>\n",
    "    <p><b><i>Given</i></b>: A collection of DNA strings of at most 1 kbp in FASTA format of equal length\n",
    "    <br><b><i>Return</i></b>: The consensus string and profile matrix for the collection</p>\n",
    "</div>\n",
    "\n",
    "#### Sample Dataset\n",
    "\n",
    "```\n",
    ">Rosalind_1\n",
    "ATCCAGCT\n",
    ">Rosalind_2\n",
    "GGGCAACT\n",
    ">Rosalind_3\n",
    "ATGGATCT\n",
    ">Rosalind_4\n",
    "AAGCAACC\n",
    ">Rosalind_5\n",
    "TTGGAACT\n",
    ">Rosalind_6\n",
    "ATGCCATT\n",
    ">Rosalind_7\n",
    "ATGGCACT\n",
    "```\n",
    "\n",
    "#### Sample Output\n",
    "\n",
    "```\n",
    "ATGCAACT\n",
    "A: 5 1 0 0 5 5 0 0\n",
    "C: 0 0 1 4 2 0 6 1\n",
    "G: 1 1 6 3 0 1 0 0\n",
    "T: 1 5 0 0 0 1 1 6\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGCAACT\n",
      "A : 5 1 0 0 5 5 0 0\n",
      "C : 0 0 1 4 2 0 6 1\n",
      "G : 1 1 6 3 0 1 0 0\n",
      "T : 1 5 0 0 0 1 1 6\n"
     ]
    }
   ],
   "source": [
    "# 1. Gather the location of the FASTA file we want to evaluate\n",
    "input_file = 'ref/rosalind_cons.txt'\n",
    "\n",
    "################################################################################\n",
    "# 2. Gather the FASTA sequences within the file by iterating through each\n",
    "#    line in the file and compiling all of the sequences under each header line\n",
    "fasta = {} \n",
    "\n",
    "current_header = \"\"\n",
    "sequence       = \"\"\n",
    "with open(input_file,'r') as file: \n",
    "    for line in file:\n",
    "        line = line.strip() # removes extra characters from a line (e.g., spaces)\n",
    "        \n",
    "        if \">\" in line: # determine if we're evaluating a header sequence \n",
    "            if len(current_header) == 0: pass \n",
    "            else:\n",
    "                # Store the previous set of sequences we've evaluated into the \n",
    "                # previous header line that we evaluated (This doesn't apply to the \n",
    "                # first header since there was nothing that came before it!)\n",
    "                fasta[current_header] = sequence.upper() # uniform case\n",
    "                \n",
    "            # Reset the state for tracking what sequence and header we're evaluating\n",
    "            current_header = line \n",
    "            sequence       = \"\"\n",
    "            fasta[line]    = \"\"\n",
    "        else: \n",
    "            # Compile all of the lines that come after the header sequence we're \n",
    "            # currently evaluating (stored under current_header)\n",
    "            sequence += line\n",
    "\n",
    "    # Add the last entry for the FASTA dictionary before we close the file\n",
    "    fasta[current_header] = sequence.upper()\n",
    "\n",
    "################################################################################\n",
    "# 4. Tally the type of nucleotides we have at each sequence position \n",
    "profile = { 'A':[], 'C':[], 'G':[], 'T':[] }\n",
    "\n",
    "for sequence in fasta:\n",
    "    sequence_to_evaluate = fasta[sequence]\n",
    "    if len(profile['A']) == 0: # initialize the profile matrix\n",
    "        profile['A'] = [0] * len(sequence_to_evaluate)\n",
    "        profile['C'] = [0] * len(sequence_to_evaluate)\n",
    "        profile['G'] = [0] * len(sequence_to_evaluate)\n",
    "        profile['T'] = [0] * len(sequence_to_evaluate)\n",
    "        \n",
    "    # Go through each sequence and each position in each sequence and tally\n",
    "    # the number of times we see a particular nucleotide at a given position\n",
    "    for i in range(len(sequence_to_evaluate)):\n",
    "        profile[sequence_to_evaluate[i]][i] += 1\n",
    "\n",
    "################################################################################\n",
    "# 5. Determine the consensus sequence\n",
    "consensus_sequence = \"\"\n",
    "\n",
    "characters = \"ACGT\"\n",
    "for i in range(len(sequence_to_evaluate)):\n",
    "    # Determine the number of A's, C's, G's, and T's we see at the current \n",
    "    # position we're evaluating, then determine which base we see the most \n",
    "    values = [profile['A'][i], profile['C'][i], profile['G'][i], profile['T'][i]]\n",
    "    consensus_sequence += characters[values.index(max(values))]\n",
    "\n",
    "################################################################################\n",
    "# 6. Show the result in the format specified by Rosalind\n",
    "print(consensus_sequence)\n",
    "for base in profile: \n",
    "    print(base,\":\",\" \".join([str(i) for i in profile[base]]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "### [Open Reading Frames](http://rosalind.info/problems/orf/)\n",
    "\n",
    "<div class=\"alert alert-block alert-success\">\n",
    "    <p><b>Exercise:</b></p>\n",
    "    <p><b><i>Given</i></b>: A DNA string, <code>s</code>, of at most 1 kbp in FASTA format\n",
    "    <br><b><i>Return</i></b>: Every distinct candidate protein string that can be translated from ORFs of <code>s</code>. Strings can be returned in any order.</p>\n",
    "</div>\n",
    "\n",
    "#### Sample Dataset\n",
    "\n",
    "```\n",
    ">Rosalind_99\n",
    "AGCCATGTAGCTAACTCAGGTTACATGGGGATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAATAAGCCTGAATGATCCGAGTAGCATCTCAG\n",
    "```\n",
    "\n",
    "#### Sample Output\n",
    "\n",
    "```\n",
    "MLLGSFRLIPKETLIQVAGSSPCNLS\n",
    "M\n",
    "MGMTPRLGLESLLE\n",
    "MTPRLGLESLLE\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MLLGSFRLIPKETLIQVAGSSPCNLS\n",
      "MGMTPRLGLESLLE\n",
      "MTPRLGLESLLE\n",
      "M\n"
     ]
    }
   ],
   "source": [
    "# 1. Gather the location of the file we want to evaluate then gather the sequence\n",
    "input_dna_filename = \"ref/rosalind_orf.txt\"\n",
    "\n",
    "dna_sequence = \"\"\n",
    "with open(input_dna_filename, \"r\") as file:\n",
    "    for line in file:\n",
    "        # Skip over the header line within our FASTA file \n",
    "        if \">\" in line: \n",
    "            continue\n",
    "        dna_sequence += line.strip().upper() # gather all lines in the FASTA \n",
    "\n",
    "################################################################################\n",
    "# 2. Determine the reverse complement of the DNA sequence \n",
    "base_complements = {\"A\":\"T\", \"T\":\"A\", \"C\":\"G\", \"G\":\"C\"}\n",
    "revcomp_dna_sequence = ''.join([base_complements[base] for base in dna_sequence][::-1])\n",
    "\n",
    "################################################################################\n",
    "# 3. Import the RNA codon table and convert it to a dictionary \n",
    "rna_codon_table_filename = \"ref/rna_codon_table.txt\"\n",
    "\n",
    "codon_table = {}\n",
    "with open(rna_codon_table_filename, \"r\") as file:\n",
    "    for line in file:\n",
    "        line = line.strip().split()\n",
    "        for codon,aa in zip([line[e] for e in range(len(line)) if e % 2 == 0],\n",
    "                            [line[e] for e in range(len(line)) if e % 2 != 0]):\n",
    "            codon_table[codon.upper()] = aa.upper() # set uniform case\n",
    "            \n",
    "################################################################################\n",
    "# 4. Determine all sets of codons that we can see from each reading frame\n",
    "#    for the sequence in the forward and reverse orientation\n",
    "\n",
    "protein_sequences = set()\n",
    "for reading_frame in [0,1,2]: \n",
    "    # Evaluate sequences in the forward orientation\n",
    "    start_codon_found = False\n",
    "    current_sequence = \"\"\n",
    "    for codon in [dna_sequence[i:i+3].upper() \n",
    "                  for i in range(reading_frame, len(dna_sequence), 3)]:\n",
    "        if len(codon) != 3:\n",
    "            break\n",
    "        codon = codon.replace(\"T\",\"U\")\n",
    "        aa_translation = codon_table[codon]\n",
    "        if aa_translation == \"M\":\n",
    "            start_codon_found = True\n",
    "            current_sequence += aa_translation\n",
    "        elif aa_translation == \"STOP\":\n",
    "            protein_sequences.add(current_sequence)\n",
    "            start_codon_found = False\n",
    "            current_sequence = \"\"\n",
    "        elif start_codon_found == True:\n",
    "            current_sequence += aa_translation\n",
    "    \n",
    "    # Evaluate sequences in the reverse orientation\n",
    "    start_codon_found = False\n",
    "    current_sequence = \"\"        \n",
    "    for codon in [revcomp_dna_sequence[i:i+3].upper() \n",
    "                  for i in range(reading_frame, len(revcomp_dna_sequence), 3)]:\n",
    "        if len(codon) != 3:\n",
    "            break\n",
    "        codon = codon.replace(\"T\",\"U\")\n",
    "        aa_translation = codon_table[codon]\n",
    "        if aa_translation == \"M\":\n",
    "            start_codon_found = True\n",
    "            current_sequence += aa_translation\n",
    "        elif aa_translation == \"STOP\":\n",
    "            protein_sequences.add(current_sequence)\n",
    "            start_codon_found = False\n",
    "            current_sequence = \"\"\n",
    "        elif start_codon_found == True:\n",
    "            current_sequence += aa_translation\n",
    "            \n",
    "################################################################################\n",
    "# 5. Double check that there aren't subsequences present within our answers\n",
    "final_protein_sequences = []\n",
    "for sequence in protein_sequences:\n",
    "    if len(sequence) == 0:\n",
    "        continue\n",
    "    # Determine where we see alternate positions to start the protein string\n",
    "    positions_of_m = [i for i, ltr in enumerate(sequence) if ltr == \"M\"] \n",
    "    if len(positions_of_m) == 1:\n",
    "        final_protein_sequences.append(sequence)\n",
    "    else:\n",
    "        for position in positions_of_m:\n",
    "            final_protein_sequences.append(sequence[position:])\n",
    "\n",
    "################################################################################\n",
    "# 6. Show the result in the format specified by Rosalind\n",
    "for sequence in final_protein_sequences:\n",
    "    print(sequence)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
